### PR TITLE
Add GLTF example with GLTF-loader working

### PR DESCRIPTION
**Pull request by georgios-uber**

### CODE REVIEW

File: examples/gltf/app.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+export const animationLoopOptions = {
+
+  onInitialize: ({gl}) => {
+
+    const gltfParser = new GLTFParser();
+    fetch(GLTF_URL).then(res => res.arrayBuffer()).then(data => {
+      gltfParser.parse(data, {createImages: false}); // see https://github.com/uber-web/loaders.gl/issues/46
+      const instantiator = new GLTFInstantiator(gl, {
+        getBuffer: accessor => {
+          // I hope Ib can save us from this terrible hack!
```
Comment (georgios-uber) [inline examples/gltf/app.js]: @ibgreen do you have a better way to access the unpackedBuffers?

File: modules/core/src/scenegraph/scenegraph-node.js [hunk 2]
```diff
@@ -40,6 +40,10 @@ export default class ScenegraphNode {
     return this;
   }
 
+  setMatrix(matrixArray) {
+    this.matrix.copy(matrixArray);
```
[the remark this answers is not in the document]
Comment (georgios-uber) [reply modules/core/src/scenegraph/scenegraph-node.js]: I prefer it that way because the **Matrix** is mutable and can result in hard-to-debug issues.

[the remark this answers is not in the document]
Comment (ibgreen) [reply modules/core/src/scenegraph/scenegraph-node.js]: > I prefer it that way because the Matrix is mutable and can result in hard-to-debug issues.

True, though the perf hit can be very significant, and not giving the user the option to do it performantly is not good.

One way out is to add a second argument that controls whether to copy or set, or just add a separate method `copyMatrix`.

File: examples/gltf/app.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    const models = [];
+
+    const gltfParser = new GLTFParser();
+    fetch(GLTF_URL).then(res => res.arrayBuffer()).then(data => {
+      gltfParser.parse(data);
+      const instantiator = new GLTFInstantiator(gl, {
+        getImage: index => {
+          const img = gltfParser.getImage(index).image;
+          const promise = new Promise(resolve => {
+            img.onload = () => {
```
Comment (ibgreen) [inline examples/gltf/app.js]: Nits: skip variable and braces: `return new Promise(resolve => resolve(...`

File: examples/gltf/app.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+export const animationLoopOptions = {
+
+  onInitialize: ({gl}) => {
+    const models = [];
+
+    const gltfParser = new GLTFParser();
+    fetch(GLTF_URL).then(res => res.arrayBuffer()).then(data => {
+      gltfParser.parse(data);
+      const instantiator = new GLTFInstantiator(gl, {
+        getImage: index => {
```
Comment (ibgreen) [inline examples/gltf/app.js]: Shouldn't this image handling be built in to the GLTFInstantiator, rather than in the example?

File: examples/gltf/app.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+              resolve(new Texture2D(gl, { data: img, format: GL.RGB }));
+            };
+          });
+
+          return promise;
+        }
+      });
+      const scene = gltfParser.getScene(0);
+      const lumaScene = instantiator.createScene(gltfParser.getScene(0));
+      console.log("gltfParser.getScene(0) = ", scene);
```
Comment (ibgreen) [inline examples/gltf/app.js]: I think we can comment out or drop this logging?

File: examples/gltf/app.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      console.log("gltfParser = ", gltfParser);
+      console.log("lumaScene = ", lumaScene);
+
+      lumaScene.traverse(node => {
+        console.log("N=", node);
```
Comment (ibgreen) [inline examples/gltf/app.js]: Ditto

File: examples/gltf/app.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+  onRender: ({gl, tick, width, height, aspect, models}) => {
+    gl.viewport(0, 0, width, height);
+    clear(gl, {color: [0, 0, 0, 1], depth: true});
+
+    const camView = new Matrix4().lookAt({eye: [0, 0.8, 1], center: [0, 0, 0], up: [0, 1, 0]}).rotateXYZ([0, tick * 0.01, 0]);
+    const camProj = new Matrix4().perspective({fov: radians(5), aspect, near: 0.01, far: 5000});
+
+    models.forEach(model => {
+      // console.log("RENDER", model);
```
Comment (ibgreen) [inline examples/gltf/app.js]: Cut

File: examples/gltf/app.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+      models
+    };
+  },
+
+  onRender: ({gl, tick, width, height, aspect, models}) => {
+    gl.viewport(0, 0, width, height);
+    clear(gl, {color: [0, 0, 0, 1], depth: true});
+
+    const camView = new Matrix4().lookAt({eye: [0, 0.8, 1], center: [0, 0, 0], up: [0, 1, 0]}).rotateXYZ([0, tick * 0.01, 0]);
+    const camProj = new Matrix4().perspective({fov: radians(5), aspect, near: 0.01, far: 5000});
```
Comment (ibgreen) [inline examples/gltf/app.js]: Nit: We don't use abbreviated names in the code base.

File: examples/pbr/controls.js [hunk 9]
```diff
@@ -21,7 +21,7 @@ class Controls {
   resetCamera() {
     this.roll = Math.PI;
     this.pitch = 0.0;
-    this.translate = 4.0;
+    this.translate = 2.5;
```
Comment (ibgreen) [inline examples/pbr/controls.js]: Should this be a different PR?

File: modules/core/src/scenegraph/base-model.js [hunk 10]
```diff
@@ -329,6 +329,40 @@ export default class BaseModel extends ScenegraphNode {
       vs = vs || MODULAR_SHADERS.vs;
       fs = fs || MODULAR_SHADERS.fs;
 
+      vs = `
```
Comment (ibgreen) [inline modules/core/src/scenegraph/base-model.js]: Can this go into mesh-model for now?

File: modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js [hunk 11]
```diff
@@ -1,13 +1,13 @@
...
+import Model from '../model';
 
 // GLTF instantiator for luma.gl
 // Walks the parsed and resolved glTF structure and builds a luma.gl scenegraph
 export default class GLTFInstantiator {
-  constructor(gl) {
+  constructor(gl, options = {}) {
     this.gl = gl;
-    this.options = {
-      removeEmptyGroups: true
```
Comment (ibgreen) [inline modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js]: Let's define a top-level constant with default options and merge them here.

```
this.options = Object.assign({}, DEFAULT_OPTIONS, options);
```

This way we can list all options in one place.

BTW, I think avoiding building unnecessary groups makes sense, not sure why we'd cut that. Otherwise the generated scenegraph can be unnecessarily deep and "verbose".

File: modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js [hunk 12]
```diff
@@ -27,75 +27,95 @@ export default class GLTFInstantiator {
...
   createPrimitive(gltfPrimitive, i, gltfMesh) {
     const attributes = this.createAttributes(gltfPrimitive.attributes, gltfPrimitive.indices);
     // TODO - handle gltfPrimitive.material
-    return new Model(this.gl, {
+    const model = new Model(this.gl, {
       id: gltfPrimitive.name || `${gltfMesh.name || gltfMesh.id}=${i}`,
-      attributes,
-      drawMode: gltfPrimitive.mode
+      drawMode: gltfPrimitive.mode || 4,
+      vertexCount: 2046 // from indices count
```
Comment (ibgreen) [inline modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js]: Huh? We can't hardcode this...

File: modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js [hunk 13]
```diff
@@ -27,75 +27,95 @@ export default class GLTFInstantiator {
...
     });
+    model.setProps({attributes});
+    this.options.getImage(0).then(img => model.setUniforms({ tex: img }));
```
Comment (ibgreen) [inline modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js]: I think it would be nice to have a single function that resolves all promises in a loaded glTF, rather than having to wait for them one by one like this.

That way a loaded file could be processed without dealing with promises. Of course if we want to support incremental loading, we'd need to do it like this.

The promise resolving function could be extended to handle URI loading etc.

File: modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js [hunk 14]
```diff
@@ -27,75 +27,95 @@ export default class GLTFInstantiator {
...
+    }
+    console.log("attributes>>", attributes, result);
+    return result;
   }
 
   createBuffer(accessor) {
-    accessor = this.createAccessor(accessor);
-    return new Buffer(this.gl, accessor).setData();
+    const opts = this.createAccessor(accessor);
+    return new Buffer(this.gl, opts);
```
Comment (ibgreen) [inline modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js]: Some TODO here as we want to create only one `Buffer` per `bufferView`?

File: modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js [hunk 15]
```diff
@@ -27,75 +27,95 @@ export default class GLTFInstantiator {
...
   }
 
   createAccessor(accessor) {
-    return null;
-    /*
-    return new Accessor({
+    console.log("AAAA",accessor)
```
Comment (ibgreen) [inline modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js]: Cut

File: modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js [hunk 16]
```diff
@@ -27,75 +27,95 @@ export default class GLTFInstantiator {
...
+    if (accessor.bufferView.byteStride) {
+      console.log("bufferView has byteStride, see https://github.com/uber-web/loaders.gl/issues/45");
+    }
+
+    return {
       type: accessor.componentType,
-      size: accessor.components,
+      // size: accessor.count, ????
```
Comment (ibgreen) [inline modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js]: Cut?

File: modules/core/src/scenegraph/group.js [hunk 17]
```diff
@@ -47,7 +47,7 @@ export default class Group extends Node {
       if (child instanceof Group) {
         result = child.traverse(visitor, {modelMatrix});
       } else {
-        // child.setUniforms({modelMatrix});
+        child.setMatrix(modelMatrix); // TODO: can child has his own matrix?
```
Comment (ibgreen) [inline modules/core/src/scenegraph/group.js]: Yes every scenegraph node has its own matrix. What we need to do is to multiply them together as we descend and pass the current matrix to `traverse(child, parentMatrix)`

File: modules/core/src/scenegraph/mesh-model.js [hunk 18]
```diff
@@ -426,6 +426,32 @@ export default class MeshModel extends Node {
       vs = vs || MODULAR_SHADERS.vs;
       fs = fs || MODULAR_SHADERS.fs;
 
+      vs = `
```
Comment (ibgreen) [inline modules/core/src/scenegraph/mesh-model.js]: Add TODO to remove this

File: modules/core/src/scenegraph/model.js [hunk 19]
```diff
@@ -217,7 +217,7 @@ export default class Model extends BaseModel {
 
   // PRIVATE METHODS
 
-  _createProgram({
+  _createProgramXXX({
```
Comment (ibgreen) [inline modules/core/src/scenegraph/model.js]: Undo

File: examples/gltf/app.js (hunk 4 above)
Comment (georgios-uber) [reply examples/gltf/app.js]: So what should the parameters of GLTFInstantiator be?

File: examples/gltf/app.js (hunk 8 above)
Comment (georgios-uber) [reply examples/gltf/app.js]: I copied this from the shadow example...

File: examples/pbr/controls.js (hunk 9 above)
Comment (georgios-uber) [reply examples/pbr/controls.js]: OK

File: modules/core/src/scenegraph/base-model.js (hunk 10 above)
Comment (georgios-uber) [reply modules/core/src/scenegraph/base-model.js]: Need to decide if we will have a dedicated shader for glTF.

File: modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js (hunk 11 above)
Comment (georgios-uber) [reply modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js]: I think we will have more edge-cases and will be harder to debug.

File: examples/gltf/app.js (hunk 4 above)
Comment (ibgreen) [reply examples/gltf/app.js]: > So what should the parameters of GLTFInstantiator be?

I guess we'll discover that as we complete the implementation. In general, the fewer options the better.

File: modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js (hunk 11 above)
Comment (ibgreen) [reply modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js]: > I think we will have more edge-cases and will be harder to debug.

Well, just to explain where I am coming from, I noticed is that in many cases we'll just end up wrapping a `Mesh` (or `MeshModel` or ...) in a `Group`. 

```
Group
  | --------|
Group     Group
  |         | 
Mesh       Mesh
```

This looked unnecessary to me, and it seemed to me that simplifying this by dropping (i.e. not generating) the middle layer of groups shouldn't make debugging harder.

My intention was not to do any complex scenegraph transformations but just be a little efficient in avoiding generation of unnecessary nodes.

File: modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js (hunk 13 above)
Comment (georgios-uber) [reply modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js]: But then I have to change the loader.
